### DISCUSSION TITLE
Support `no-cache` execution

### DISCUSSION THREAD

**Discussion by sipsma** [dagger/dagger#12063] [Ideas]
There are use cases that call for always running a certain step of a pipeline. Examples include an step that checks to see if a remote resource has drifted from the last known cached state, in which case other steps may need to run in order to correct it.

This requires being able to specify that a step should always run. A few open questions:
1. Who configures this: the extension author (or us as dagger devs in the case of core api)? the api invoker? both?
2. Do we use buildkit's `IgnoreCache` metadata to implement this (and accept the baggage of, i.e. cache mounts being empty) or do we use the "random env" approach (and accept the baggage of creating cache entries in buildkit that will never be reused)?
   * Or do we do some upstream work to implement a better approach than those two?

**Comment by jpadams** [comment]
https://discord.com/channels/707636530424053791/1042823572000886854/1042823572000886854
https://discord.com/channels/707636530424053791/1042418867487645736/1042418867487645736
https://discord.com/channels/707636530424053791/1040899837232681003/1041222646576189560
🔥

**Comment by sipsma** [comment]
Some updated thoughts on this, another benefit to having official support for this is in our API that we can have our remote caching skip these execs (which would obviously be a waste of time+space to cache).

I think a good potential implementation here would be to re-use @vito's idea w/ services where we stop buildkit from caching these by forcing them to fail. In this case we wouldn't cancel them though, we can instead have the shim always exit non-zero so buildkit thinks it's failed and thus won't cache it. We store the real exit code in the meta mount already, so we'll just use that to return the "real" error or no error if it's 0.

This would avoid the problems with buildkit's default `IgnoreCache` setting which results in cache mounts being empty while also avoiding unneeded saving of layers in the local and remote cache.
